### PR TITLE
Adding allowed_categories option to graphql_cowboy_handler

### PR DESCRIPTION
**Pull request by JanuszJakubiec**
Adding new config option for graphql_cowboy_handler allowing to insert the list of enabled categories.
When the list is empty, which is by default, all categories are enabled. Otherwise, when a request from a category that is not in a list comes, it will be rejected with the error "category disabled".

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146279) / small_tests / 4db0b32fabb6e8a40c99a6778f891f9fa3f94a45
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/146279/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/146279/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-24_07.31.50/lib.mongooseim.logs/run.2022-10-24_07.31.51/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146353) / small_tests / 0af69a99f68271d89caf002246aa6391abd4c78a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/146353/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/146353/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-24_09.48.29/lib.mongooseim.logs/run.2022-10-24_09.48.30/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/146931) / small_tests / 12e20913a63c7bbb94cabfaeeed8b3dee60da81c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/146931/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/146931/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-25_10.39.35/lib.mongooseim.logs/run.2022-10-25_10.39.36/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148558) / small_tests / bf8983e603fff4abe296667a194b7daca726f42a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/148558/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/148558/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-03_09.56.21/lib.mongooseim.logs/run.2022-11-03_09.56.22/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148587) / small_tests / ff312860d36c602ac74a233ffe2f4c8fb64e92f9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/148587/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/148587/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-03_10.03.32/lib.mongooseim.logs/run.2022-11-03_10.03.33/suite.log.html)

**Review by chrzaszcz — COMMENTED** [review]
The documentation for the new configuration field is missing.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149836) / small_tests / 5b3eb1d2c3bdf23ea00e4ba4e4e3638081147ae4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/149836/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/149836/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_08.08.31/lib.mongooseim.logs/run.2022-11-08_08.08.31/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150322) / small_tests / 4d4a0e039a1bd329d218931e19b241b6e9b3641a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/150322/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/150322/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_17.42.09/lib.mongooseim.logs/run.2022-11-08_17.42.10/suite.log.html)

File: src/graphql/mongoose_graphql_check_categories.erl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            Document;
+        _ ->
+            Definitions2 = lists:map(fun(#op{schema = Schema} = Op) ->
+                parse_schema(Schema, Op, Categories)
+            end, Definitions),
+            #document{definitions = Definitions2}
+    end.
+
+parse_schema(#object_type{fields = Fields} = Schema, Op, Categories) ->
+    Fields2 = lists:foldl(fun({Key, Value}, Acc) ->
```
Comment (chrzaszcz) [inline src/graphql/mongoose_graphql_check_categories.erl:28]: Just a note: using `maps:map` would be cleaner here.

File: big_tests/tests/graphql_SUITE.erl [hunk 2]
```diff
@@ -136,6 +159,32 @@ auth_domain_admin_checks_auth(Config) ->
     Res = execute_auth(admin_check_auth_body(), Config),
     ?assertAdminAuth(Domain, 'DOMAIN_ADMIN', 'AUTHORIZED', Res).
 
+category_disabled_error_test(Config) ->
+    Status = execute_auth(admin_server_get_loglevel_body(), Config),
+    {_Code, #{<<"errors">> := [Msg]}} = Status,
+    ?assertEqual(<<"category_disabled">>, get_value([extensions, code], Msg)),
+    ?assertEqual([<<"server">>], get_value([path], Msg)).
+
+category_does_not_exists_error(Config) ->
```
Comment (chrzaszcz) [inline big_tests/tests/graphql_SUITE.erl:168]:
```suggestion
category_does_not_exist_error(Config) ->
```

**Review by chrzaszcz — COMMENTED** [review]
The docs are still missing, see previous code review.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150507) / small_tests / f188f9d6d8f6e8ba70900283caf88e6e6b9a19e1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/150507/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/150507/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-09_12.43.51/lib.mongooseim.logs/run.2022-11-09_12.43.51/suite.log.html)

File: doc/configuration/listen.md [hunk 3]
```diff
@@ -521,7 +521,13 @@ When set, enables authentication for the admin API, otherwise it is disabled. Re
 * **Default:** not set
 * **Example:** `password = "secret"`
 
-Required to enable authentication for the admin API.
+#### `listen.http.handlers.mongoose_graphql_cowboy_handler.allowed_categories`
+* **Syntax:** non-empty array of strings. Allowed values: `"checkAuth", "account", "domain", "last", "muc", "muc_light", "session", "stanza", "roster", "vcard", "private", "metric", "stat", "gdpr", "mnesia", "server", "inbox", "http_upload", "offline", "token"`
+* **Default:** all graphQL categories enabled
+* **Example:** `allowed_categories = ["domain", "last"]`
+
+By default, when the option is not included, all graphQL categories are enabled, so you don't need to add this option.
```
Comment (chrzaszcz) [inline doc/configuration/listen.md:529]:
```suggestion
By default, when the option is not included, all GraphQL categories are enabled, so you don't need to add this option.
```

File: doc/configuration/listen.md [hunk 4]
```diff
@@ -521,7 +521,13 @@ When set, enables authentication for the admin API, otherwise it is disabled. Re
...
+When this option is added, only listed graphQL categories will be processed. For others, the error "category disabled" will be returned.
```
Comment (chrzaszcz) [inline doc/configuration/listen.md:530]:
```suggestion
When this option is added, only listed GraphQL categories will be processed. For others, the error "category disabled" will be returned.
```

**Review by chrzaszcz — APPROVED** [review]
Looks good :+1:

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150821) / small_tests / c1dd1d77976cedf7dcc3bf9a7f9b015c28660376
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3822/150821/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3822/150821/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-10_07.20.11/lib.mongooseim.logs/run.2022-11-10_07.20.11/suite.log.html)

**Pull request merged**